<launch>
    <param name="/use_sim_time" value="true"/>

    <node pkg="robot_localization" type="ukf_localization_node" name="ekf_se" clear_params="true">
        <rosparam command="load" file="$(find pure_pursuit)/config/ekf_params.yaml" />
    </node>

    <node pkg="amcl" type="amcl" name="amcl" output="screen">
       <rosparam command="load" file="$(find pure_pursuit)/config/amcl_params.yaml" />
    </node>

    <node name="map_server" pkg="map_server" type="map_server" args="$(find teb1)/map/map.yaml" />

    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
        <rosparam file="$(find pure_pursuit)/config/costmap_common_params.yaml" command="load" ns="global_costmap" />
        <rosparam file="$(find pure_pursuit)/config/global_costmap_params.yaml" command="load" />

        <param name="base_global_planner" value="global_planner/GlobalPlanner" />
        <param name="planner_frequency" value="10.0" />
        <param name="planner_patience" value="10.0" />
        
        <param name="controller_frequency" value="0.001" />

        <param name="clearing_rotation_allowed" value="false" />

        <remap from="/cmd_vel" to="/teb/cmd_vel" />
    </node>

    <node pkg="pure_pursuit" name="pure_pursuit_controller" type="controller" respawn="false" output="screen">
        
        <param name="controller_freq" value="40" />

        <!-- speed -->
        <param name="can_speed_change" value="true" />
        <param name="baseSpeed" value="2.5" />
        <param name="min_speed" value="1.5" />
        <param name="max_speed" value="10.0" />
        <param name="rush_speed" value="5.0" />
        <param name="rush_dist" value="6.0" />
        <param name="acc_limit" value="0.15" />
        <param name="acc_weight" value="2.0" />
        <param name="back_weight" value="1.0" />

        <!-- angle -->
        <param name="baseAngle" value="0" />

        <!--  eta filter -->
        <param name="eta_weight" value="0.9" />

        <!-- pid -->
        <param name="kp" value="0.36" /> <!-- 1.36 -->
        <param name="kd" value="4.5" />  <!-- 4.5 -->
        
        <!-- Lfw -->
        <param name="Lfw" value="0.8" />
        <param name="can_Lfw_change" value="false" />
        <param name="Lfw_min" value="0.6" />
        <param name="Lfw_max" value="1.0" />

        <!-- back -->
        <param name="can_back" value="false" />
        <param name="back_speed" value="-3.0" />
        <param name="can_back_turn" value="false" />

        <remap from="/car/cmd_vel" to="/vesc_sim/cmd_vel" />

    </node>

    <node name="rviz" pkg="rviz" type="rviz"
          args="-d $(find pure_pursuit)/config/pure_pursuit.rviz" required="true"/>

</launch>